### PR TITLE
Use pylibcudf contiguous split APIs in cudf python

### PR DESCRIPTION
**Pull request by Matt711**
## Description

Apart of #15162 
## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/copying.pyx [hunk 1]
```diff
@@ -392,35 +375,31 @@ cdef class _CPackedColumns:
...
         header["column-names"] = self.column_names
         header["index-names"] = self.index_names
-        if self.c_obj.metadata.get()[0].data() != NULL:
+        cdef PlcPackedColumns p = self._data
+        if p.c_obj.get()[0].metadata.get()[0].data() != NULL:
             header["metadata"] = list(
-                <uint8_t[:self.c_obj.metadata.get()[0].size()]>
-                self.c_obj.metadata.get()[0].data()
+                <uint8_t[:p.c_obj.get()[0].metadata.get()[0].size()]>
+                p.c_obj.get()[0].metadata.get()[0].data()
```
Comment (wence-) [inline python/cudf/cudf/_lib/copying.pyx:382]: And I think we can simplify all of this to send `self._metadata` and `self._gpudata` directly.

File: python/pylibcudf/pylibcudf/contiguous_split.pyx [hunk 2]
```diff
@@ -87,6 +88,33 @@ cdef class PackedColumns:
...
+    @property
+    def gpu_data_ptr(self):
+        if self.c_obj.get() != NULL:
+            return int(<uintptr_t>self.c_obj.get()[0].gpu_data.get()[0].data())
+
+    @property
+    def gpu_data_size(self):
+        if self.c_obj.get() != NULL:
+            return int(<size_t>self.c_obj.get()[0].gpu_data.get()[0].size())
+
```
[the remark this answers is not in the document]
Comment (Matt711) [reply python/pylibcudf/pylibcudf/contiguous_split.pyx]: Done.

File: python/cudf/cudf/_lib/copying.pyx [hunk 3]
```diff
@@ -429,28 +408,64 @@ cdef class _CPackedColumns:
...
+                column_dtypes[name] = col.dtype
+
+        return cls(
+            plc.contiguous_split.PackedColumns.from_plc_table(
+                plc.Table(
+                    [
+                        col.to_pylibcudf(mode="read") for col in columns
+                    ]
+                )
+            ),
```
[the remark this answers is not in the document]
Comment (Matt711) [reply python/cudf/cudf/_lib/copying.pyx]: Yep, thanks

File: python/cudf/cudf/_lib/copying.pyx [hunk 4]
```diff
@@ -331,54 +326,32 @@ def get_element(Column input_column, size_type index):
...
-    def gpu_data_size(self):
-        return int(<size_t>self.c_obj.gpu_data.get()[0].size())
+    def __cuda_array_interface__(self):
+        return self._gpu_data.__cuda_array_interface__
 
     def serialize(self):
         header = {}
         frames = []
-
+        cdef DeviceBuffer dbuf = self._gpu_data.obj
```
Comment (Matt711) [inline python/cudf/cudf/_lib/copying.pyx:351]:
```suggestion
```

File: python/cudf/cudf/_lib/copying.pyx [hunk 5]
```diff
@@ -331,54 +326,32 @@ def get_element(Column input_column, size_type index):
...
         gpu_data = as_buffer(
-            data=self.gpu_data_ptr,
-            size=self.gpu_data_size,
+            data = int(<uintptr_t>dbuf.c_obj.get()[0].data()),
+            size = int(<uintptr_t>dbuf.c_obj.get()[0].size()),
```
Comment (Matt711) [inline python/cudf/cudf/_lib/copying.pyx:354]:
```suggestion
            data = self._gpu_data.ptr,
            size = self._gpu_data.size,
```

**Review by Matt711 — COMMENTED** [review]
simplify

**Review by wence- — APPROVED** [review]
Small suggestion to avoid needing `setstate/getstate` on `HostBuffer`, otherwise looks good, thanks.

File: python/pylibcudf/pylibcudf/contiguous_split.pyx [hunk 6]
```diff
@@ -63,6 +63,23 @@ cdef class HostBuffer:
...
+        new_c_obj.get().reserve(len(state))
+        for byte in state:
+            new_c_obj.get().push_back(<uint8_t>byte)
+
+        self.c_obj = move(new_c_obj)
+        self.nbytes = dereference(self.c_obj).size()
+        self.shape[0] = self.nbytes
+        self.strides[0] = 1
+
+
```
[the remark this answers is not in the document]
Comment (Matt711) [reply python/pylibcudf/pylibcudf/contiguous_split.pyx]: Thanks for the simplification

**Pull request merged**